Remove index mapper dynamic settings

PnPie opened the pull request:
Remove `index.mapper.dynamic` setting.
Relates to #25719

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jpountz commented:
@PnPie Ideally we would only remove this setting for 6.0 indices, and keep it working for 5.x indices.

dakrone commented:
@jpountz you self-requested a review, is there more reviewing for this or does it need to be changed for working with 6.0 indices and removed for 7.0 indices as you commented?

jpountz commented:
@dakrone It needs to be changed to take the creation version into account indeed.

dakrone commented:
@PnPie would you be able to make the changes @jpountz requested?

PnPie commented:
@jpountz @dakrone srry for the late response, I didn't well understand that @jpountz mentioned in fact, could you please explain a bit more about the modification and what does it mean "to take the creation version into account" ? I'm on vacation these days and maybe I cannot change it very quickly.

jpountz commented:
@PnPie We would like to keep this feature working for indices that have been created with Elasticsearch 5.x. So the idea would be to undo your changes completely, replace

```java
this.dynamic = this.indexSettings.getValue(INDEX_MAPPER_DYNAMIC_SETTING);
```
with something like
```java
if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_6_0_0)) {
    if (INDEX_MAPPER_DYNAMIC_SETTING.exists(indexSettings.getSettings())) {
        throw new IllegalArgumentException("Setting [" + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + "] is removed as of 6.0.0 since it became useless with the removal of types");
    }
    this.dynamic = INDEX_MAPPER_DYNAMIC_DEFAULT;
} else {
    this.dynamic = this.indexSettings.getValue(INDEX_MAPPER_DYNAMIC_SETTING);
}
```
Then tests need to be fixed in order to set the creation version explicitly so that they can keep testing this functionality. And ideally we'd need a new test that makes sure that we fail index creation if this setting is set.

PnPie commented:
@jpountz I did the modification as mentioned and I removed 2 test cases for the disability of `index.mapper.dynamic" at cluster level as it is already removed in current version (and it seems setting index version explicitly for these 2 cases is a little bit complicated ?). I also added a version test class to test the setting in current and 5.0 version.

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -148,7 +148,14 @@ public MapperService(IndexSettings indexSettings, IndexAnalyzers indexAnalyzers,
+            this.dynamic = INDEX_MAPPER_DYNAMIC_DEFAULT;
+        } else {
+            this.dynamic = this.indexSettings.getValue(INDEX_MAPPER_DYNAMIC_SETTING);
+        }
```
master is the future 7.0, so I would do the following:
```java
if (INDEX_MAPPER_DYNAMIC_SETTING.exists(indexSettings.getSettings())) {
    if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_7_0_0)) {
        // 7.x index
        throw new IllegalArgumentException("Setting " + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + " was removed after version 6.0.0");
    } else {
        // 6.x index
        DEPRECATION_LOGGER.deprecated("Setting " + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + " is deprecated since indices may not have more than one type anymore.");
    }
}
```

Then when backporting I'll just remove the 7.x branch and make sure that we only emit a deprecation warning on 6.x indices (you don't need to worry about it).

jpountz commented:
@elasticmachine please test it

jpountz commented:
@elasticmachine please test it

jpountz commented:
The failing test passes when I rebase. I'll merge. Thanks @PnPie !

javanna commented:
@jpountz this is a breaking change right? Would you mind adding labels? Thanks!
